src/crimson/os/seastore/seastore_types.h

```diff
@@ -1025,7 +1026,120 @@ inline extent_len_le_t init_extent_len_le(extent_len_t len) {
+ * Besides the basic properties inherits from integers, such as strong ordering
+ * and arithmetic operations, some types of laddr_t also includes the properties
+ * derived from user data. The static layout design ensures these properties are
+ * deterministic and predictable, allowing more optimization based on laddr_t.
+ *
+ * # Overview
+ * The layout of laddr_t has three parts:
+ *
+ * [upgrade:1][object_info:79][object_content:48]
+ *
+ * Each pair of square brackets represents a property stored within the laddr_t
+ * interger. Each property contains a name and the count of bits it used.
+ *
+ * When the object info for different objects does not conflict, we obtain a
+ * useful property:
+ * Each object could have a unique laddr prefix within an OSD.
+ *
+ * Base on this property, we could:
+ * 1. Group different snapshots of on object to the same laddr prefix, which
+ *    could speed up cloning process.
+ * 2. Track frequently accessed objects using laddr_t without keeping the full
+ *    object name.
+ * 3. Effectively remove extents of an object via range deletion.
+ *
+ * # Object Info
+ * The defination of this property is:
+ * [pool:12][shard:8][reverse_hash:32][local_object_id:27]
+ *
+ * The pool, shard and reverse_hash come from the information of object;
+ * The local_object_id is just a random number to handle the collision between
+ * objects. When a collision occurs, the local clone ids of conflicting objects
+ * should be different.
+ *
+ * For global meta logical extents, such as RootBlock, CollectionNode and
+ * SeastoreNodeExtent, these fields should be set to zero, and other objects
+ * data should never use this prefix.
+ *
+ * # Object Content
+ * TODO: This property might change at runtime.
+ *
+ * The defination of this property is:
+ * [is_metadata:1][object_data:47]
+ *
+ * The `is_metadata` field is fixed and always located at the most significant
+ * bit of the object content, so it can't be changed at runtime.
+ *
+ * ## is_metadata true
+ * The remaining bits represent the address of OMap*Node. Take a random value
+ * as address when allocating a new omap extent.
+ *
+ * TODO: It's possiable for OMapInnerNode store only the value of this field
+ * instead of a full 128-bits laddr_t to increase the fan-out of the omap node.
+ *
+ * ## is_metadata false
+ * The remaining bits represent the address of ObjectDataBlock. It can be
+ * further divied into the following layout, which could also change at runtime:
+ *
+ * [local_clone_id:x][data_offset:47-x]
+ *
+ * local_clone_id is used when taking snapshots for objects from RBD and CephFS
+ * pools.
+ * data_offset is the address of extents, it is block aligned and represents a
+ * 4KiB block on disk.
+ *
+ * When creating a new snapshot, taking a random local clone id as the new
+ * data base for snap object. Different snaps has different local clone ids
+ * and should not conflict with other snaps.
```

Comment:
Forget to mention here how is the value of indirect mapping optimized? I suppose its value only need to store x-bits of `local_clone_id`?

Reply:
Yeah, will add this case.